packages/expect/src/index.js

```diff
@@ -307,9 +313,13 @@ const makeThrowingMatcher = (
 
       if (isPromise((potentialResult: any))) {
         const asyncResult = ((potentialResult: any): AsyncExpectationResult);
+        const asyncError = new JestAssertionError();
+        if (Error.captureStackTrace) {
+          Error.captureStackTrace(asyncError, throwingMatcher);
+        }
 
         return asyncResult
-          .then(aResult => processResult(aResult))
+          .then(aResult => processResult(aResult, asyncError))
```

Comment:
I'm having some issues with this code a bit further on. I'm almost certainly not understanding correctly, but it seems to me....

1) `asyncError` is always truthy here. Even though this is the normal codepath that any async matcher will follow

2) In `processResult` we just test the truthiness of `asyncError` to decide whether an error was thrown. 

3) That means we end up `throw`ing a normal `{ message, pass }` assertion result, which is not what is supposed to happen.

But.... what that suggests to me is that async custom would just be fully broken all the time, which.... I mean, I assume they're tested? So I assume I'm missing something?

But for me, they are in fact broken all the time.

Reply:
What issue are you having? The same one that I fixed in the description or a different one? It's been a while since I worked on this, so I don't remember what we're supposed to be throwing for point #3. Can you remind me what we're supposed to be throwing instead?

`asyncError` will indeed always be defined if we're using async matchers. However, the only time we do something with the `asyncError` is when the assertion fails because of this conditional [here](https://github.com/donald-s/jest/blob/ef15826ae38dd77cbdafe3eb56b05c496be4e4e7/packages/expect/src/index.js#L262). Perhaps I'm misunderstanding what you're saying?